File: kotlin/scon-core/src/main/kotlin/io/github/realmlabs/scon/Value.kt
package io.github.realmlabs.scon

public sealed interface SconValue {
    public data object Null : SconValue
    public data class Bool(val value: Boolean) : SconValue
    public data class Number(val value: SconNumber) : SconValue
    public data class StringValue(val value: String) : SconValue
    public data class ArrayValue(val values: List<SconValue>) : SconValue
    public data class ObjectValue(val values: LinkedHashMap<String, SconValue>) : SconValue
}

public sealed interface SconNumber {
    public fun toSconString(): String

    public data class I64(val value: Long) : SconNumber {
        override fun toSconString(): String = value.toString()
    }

    public data class U64(val value: ULong) : SconNumber {
        override fun toSconString(): String = value.toString()
    }

    public data class F64(val value: Double) : SconNumber {
        override fun toSconString(): String = value.toString()
    }

    public companion object {
        public fun parse(raw: String): SconNumber {
            return if (raw.indexOf('.') >= 0 || raw.indexOf('e') >= 0 || raw.indexOf('E') >= 0) {
                val value = raw.toDoubleOrNull()
                    ?: throw invalidNumber(raw)
                if (!value.isFinite()) throw invalidNumber(raw)
                F64(value)
            } else if (raw.startsWith("-")) {
                I64(raw.toLongOrNull() ?: throw invalidNumber(raw))
            } else {
                U64(raw.toULongOrNull() ?: throw invalidNumber(raw))
            }
        }

        private fun invalidNumber(raw: String): SconException =
            SconException(
                SconError(
                    code = SconErrorCode.InvalidNumber,
                    message = "invalid SCON number: $raw",
                ),
            )
    }

}

File: python/src/scon/value.py
from __future__ import annotations

from collections import OrderedDict
from dataclasses import dataclass
from math import isfinite
from typing import Literal

from .error import SconError

NumberKind = Literal["i64", "u64", "f64"]


@dataclass(frozen=True)
class SconNumber:
    kind: NumberKind
    value: int | float

    @staticmethod
    def parse(raw: str) -> SconNumber:
        try:
            if "." in raw or "e" in raw or "E" in raw:
                value = float(raw)
                if not isfinite(value):
                    raise ValueError("float value must be finite")
                return SconNumber("f64", value)
            value = int(raw, 10)
            if raw.startswith("-"):
                if value < -(1 << 63):
                    raise ValueError("signed integer is below i64 minimum")
                return SconNumber("i64", value)
            if value > (1 << 64) - 1:
                raise ValueError("unsigned integer exceeds u64 maximum")
            return SconNumber("u64", value)
        except ValueError as exc:
            raise SconError("InvalidNumber", f"invalid SCON number {raw!r}: {exc}") from exc

    @staticmethod
    def from_float(value: float) -> SconNumber:
        if not isfinite(value):
            raise SconError("Serde", "non-finite floats cannot be serialized")
        return SconNumber("f64", value)

    def as_i64(self) -> int:
        if self.kind == "i64":
            return int(self.value)
        if self.kind == "u64" and int(self.value) <= (1 << 63) - 1:
            return int(self.value)
        raise SconError("Serde", "integer overflow")

    def as_u64(self) -> int:
        if self.kind == "u64":
            return int(self.value)
        if self.kind == "i64" and int(self.value) >= 0:
            return int(self.value)
        raise SconError("Serde", "integer overflow")

    def as_float(self) -> float:
        return float(self.value)

    def to_text(self) -> str:
        if self.kind == "f64":
            return format(float(self.value), ".15g")
        return str(int(self.value))


class SconArray(list["SconValue"]):
    pass


class SconObject(OrderedDict[str, "SconValue"]):
    pass


SconValue = None | bool | str | SconNumber | SconArray | SconObject


def clone_value(value: SconValue) -> SconValue:
    if isinstance(value, SconArray):
        return SconArray(clone_value(item) for item in value)
    if isinstance(value, SconObject):
        out = SconObject()
        for key, item in value.items():
            out[key] = clone_value(item)
        return out
    return value
